New article "Let agents inspect w/ Chrome DevTools MCP" (#3811)
MicrosoftDocs/edge-developer: microsoft-edge/web-platform/devtools-mcp-server.md

Reviewer: We should not call this mcp.md. MCP is a pretty general term and we might want to use it for another article in the future. This page should perhaps be called devtools-mcp-server.md instead, which is more specific to this.

| Canary | `%LOCALAPPDATA%\Microsoft\Edge SxS\Application\msedge.exe` |

##### [macOS](#tab/macos/)
   
| Channel | Default Path |
|---|---|
| Stable | `/Applications/Microsoft Edge.app/Contents/MacOS/Microsoft Edge` |
| Beta | `/Applications/Microsoft Edge Beta.app/Contents/MacOS/Microsoft Edge` |
| Dev | `/Applications/Microsoft Edge Dev.app/Contents/MacOS/Microsoft Edge` |
| Canary | `/Applications/Microsoft Edge Canary.app/Contents/MacOS/Microsoft Edge` |

##### [Linux](#tab/linux/)

| Channel | Default Path |
|---|---|
| Stable | `/usr/bin/microsoft-edge` |
| Beta | `/usr/bin/microsoft-edge-beta` |
| Dev | `/usr/bin/microsoft-edge-dev` |

---

> [!NOTE]
> These are default install locations.  Paths may vary based on user configuration, version, or group policies.  Linux does not have a Canary channel.

<!-- ====================================================================== -->
## 2. Auto-connecting to a running Edge instance

In this mode, you start Edge yourself (or it's already running), and the MCP server connects to it.  This is useful when:

* You want to maintain browser state between manual testing and agent-driven testing.

> * You need to be signed into a website that blocks automated browser launches.

Reviewer: I don't understand what we mean by a website that blocks automated browser launches. How can a website blocks this?

Author: What's meant by "a website that blocks automated browser launches". How can a website block automated browser launches?

https://review.learn.microsoft.com/en-us/microsoft-edge/web-platform/devtools-mcp-server?branch=pr-en-us-3811&tabs=windows#auto-connect-to-a-running-edge-instance

@zoherghadyali

Reviewer 2: Yeah I think this should just say instead "You need to be signed into a website"

We could add additional details like "You need to be signed into a website with a complex authentication flow that's hard to accomplish with traditional web automation" but that's quite wordy

Author: <img width="1388" height="269" alt="image" src="https://github.com/user-attachments/assets/2ae900b8-95c1-4a2f-ab8e-7f6bb720f12d" />

```

<!-- ------------------------------ -->
#### Step 2: Find the WebView2 user data directory

You need to discover the user data directory of the host app.

The path for the user data directory ends with `\EBWebView`.  The `\EBWebView` suffix is automatically added by the WebView2 Runtime.  If you're copying the user data path from source code, you might need to append `\EBWebView`.

You can find the path for the user data directory by using WebView2Utilities: In the **Host Apps** tab, select your running host app, and examine the **User data folder** row.

<!-- ------------------------------ -->
#### Step 3: Configure the MCP server

```json
{
  "servers": {
    "chrome-devtools": {
      "type": "stdio",
      "command": "npx",
      "args": [
        "-y",
        "chrome-devtools-mcp@latest",
        "--autoConnect",
        "--user-data-dir=%LocalAppData%\\Packages\\<APP_PACKAGE>\\LocalState\\EBWebView"
      ]
    }
  }
}
```

Replace `<APP_PACKAGE>` with the package name of your host app.

<!-- ------------------------------ -->
#### Step 4: Test your setup

Launch your WebView2 host app, then use your MCP client to interact with your WebView2 host app:

> `Take a snapshot of the current page`<!-- todo: what is this? -->

Author: "Take a snapshot of the current page" - what is this?

https://review.learn.microsoft.com/en-us/microsoft-edge/web-platform/devtools-mcp-server?branch=pr-en-us-3811&tabs=windows#step-4-test-your-setup

@zoherghadyali

Reviewer: This is a prompt which you send to your AI coding agent.

```suggestion
Enter the following prompt in your Model-Context-Protocol (MCP) client:

`Take a snapshot of the current page`

The MCP server should connect to your WebView2 host app instance and execute the command.
```

<!-- ====================================================================== -->
## Troubleshooting

<!-- ------------------------------ -->
#### "Could not connect to Chrome" error with auto-connect

If you are using auto-connect and get an error such as "Could not connect to Chrome"<!-- todo: with Microsoft Edge? -->:

* Make sure Microsoft Edge is running, and remote debugging is enabled.
* Make sure the `--user-data-dir` path is correct for your channel of Microsoft Edge.
* Make sure the `DevToolsActivePort` file exists in the user data directory.  The `DevToolsActivePort` file is created when remote debugging is active.

<!-- ------------------------------ -->
#### Edge not found at `executablePath`

* Make sure that the `executablePath` for the channel of Microsoft Edge exists on disk.  For example, Edge Canary on Windows installs per-user under `%LOCALAPPDATA%`, not under `\Program Files\`.
* On Windows, use double backslashes (`\\`) in JSON strings.

<!-- ------------------------------ -->
#### WebView2 won't connect

* Make sure the registry key matches your host app's executable name exactly.

> * Restart the host app after adding the registry key.  The argument<!-- todo: which argument? The registry key argument? --> is only read at WebView2 creation time.

Author: "The argument [which arg?] is only read at WebView2 creation time."

https://review.learn.microsoft.com/en-us/microsoft-edge/web-platform/devtools-mcp-server?branch=pr-en-us-3811&tabs=windows#webview2-wont-connect

Author: * After adding the `AdditionalBrowserArguments` registry key, restart the host 
app.  The app name `.exe` argument is only read at WebView2 creation time.

| Beta | `~/Library/Application Support/Microsoft Edge Beta` |
| Dev | `~/Library/Application Support/Microsoft Edge Dev` |
| Canary | `~/Library/Application Support/Microsoft Edge Canary` |

##### [Linux](#tab/linux/)

| Channel | Default Path |
|---|---|
| Stable | `~/.config/microsoft-edge` |
| Beta | `~/.config/microsoft-edge-beta` |
| Dev | `~/.config/microsoft-edge-dev` |
| Canary | n/a; there isn't a Canary channel for Linux. |

---

<!-- ====================================================================== -->
## Auto-connect to a WebView2 instance

WebView2 doesn't have a "launch" scenario; instead, the host app creates the WebView2 instance.

The MCP server connects to the WebView2 instance via auto-connect, similar to [Auto-connect to a running Edge instance](#auto-connect-to-a-running-edge-instance), above.  

<!-- ------------------------------ -->
#### Step 1: Enable remote debugging for WebView2

You need to configure the WebView2 runtime to start with `--remote-debugging-port=0`.  There are two ways to do this: WebView2Utilities, or Windows Registry.

<!-- ---------- -->
###### By using WebView2Utilities

> In this approach, you use WebView2Utilities and follow the steps in [How to: Auto open DevTools](https://github.com/david-risney/WebView2Utilities/wiki/How-to:-Auto-open-DevTools).  But instead of selecting the **Auto open DevTools** checkbox, in the **Arguments** textbox, enter `--remote-debugging-port=0`.

Reviewer: @zoherghadyali would you be ok if we inlined the wiki instructions? This would avoid potential confusion/context switching. The wiki isn't that long.

Author: https://review.learn.microsoft.com/en-us/microsoft-edge/web-platform/devtools-mcp-server?branch=pr-en-us-3811&tabs=windows#by-using-webview2utilities

Reviewer 2: Sounds good to me!

Author: <img width="1651" height="849" alt="image" src="https://github.com/user-attachments/assets/9793edbd-9f35-4ff4-aee2-d44b2aa5ce19" />

<!-- ---------- -->
###### By using the Windows Registry

In this approach, you create a registry value to set `AdditionalBrowserArguments`.  Replace `appname.exe` with the name of your WebView2 host executable:

```
[HKEY_CURRENT_USER\Software\Policies\Microsoft\Edge\WebView2\AdditionalBrowserArguments]
"appname.exe"="--remote-debugging-port=0"
```

<!-- ------------------------------ -->
#### Step 2: Find the WebView2 user data directory

You need to discover the user data directory of the host app.

The path for the user data directory ends with `\EBWebView`.  The `\EBWebView` suffix is automatically added by the WebView2 Runtime.  If you're copying the user data path from source code, you might need to append `\EBWebView`.

> You can find the path for the user data directory by using WebView2Utilities: In the **Host Apps** tab, select your running host app, and examine the **User data folder** row.

Reviewer: ```suggestion
You can find the path for the user data directory by using WebView2Utilities: in the **Host Apps** tab, select your running host app, and examine the **User data folder** row.
```

<!-- ------------------------------ -->
#### Step 3: Configure the MCP server

> 

Reviewer: ```suggestion

To configure the Model-Context-Protocol (MCP) server, use `--autoConnect` combined with `--user-data-dir` pointing to the WebView2 user data directory.

```

      "type": "local",
      "command": "npx",
      "args": [
        "-y",
        "chrome-devtools-mcp@latest",
        "--executablePath=%ProgramFiles(x86)%\\Microsoft\\Edge\\Application\\msedge.exe"
      ]
    }
  }
}
```

You can also add an MCP server interactively by running `copilot` and then `/mcp add`.

<!-- ------------------------------ -->
#### Other MCP clients, such as Claude Code, Cursor, or Gemini CLI

Most Model-Context-Protocol (MCP) clients use the generic `mcpServers` format without a `type` field:

```json
{
  "mcpServers": {
    "chrome-devtools": {
      "command": "npx",
      "args": [
        "-y",
        "chrome-devtools-mcp@latest",
        "--executablePath=%ProgramFiles(x86)%\\Microsoft\\Edge\\Application\\msedge.exe"
      ]
    }
  }
}
```

The `args` array is the same across all clients; only the wrapper format differs.  See your MCP client's documentation for the exact config file location and format.

<!-- ====================================================================== -->
## Troubleshooting

<!-- ------------------------------ -->
#### "Could not connect to Chrome"<!-- todo: to Microsoft Edge? --> error with auto-connect

> If you are using auto-connect and get an error such as "Could not connect to Chrome"<!-- todo: to Microsoft Edge? -->:

Author: "Could not connect to Chrome" - to Microsoft Edge?   2x

https://review.learn.microsoft.com/en-us/microsoft-edge/web-platform/devtools-mcp-server?branch=pr-en-us-3811&tabs=windows#could-not-connect-to-chrome-error-with-auto-connect

Reviewer: I suspect this string is coming from the Chrome DevTools MCP server. We're using this server to connect to Edge or WebView2, because it's inherently compatible with them, being Chromium-based, but we don't control what's in the Chrome DevTools MCP server itself. It may very well be that the server has that string "Could not connect to Chrome" hard-coded in it.

@zoherghadyali can confirm.

Reviewer 2: That's correct. Since we're recommending the Chrome DevTools MCP server here, I think we can leave this error as "Could not connect to Chrome"

---
title: Let agents inspect your site and WebView2 app with Chrome DevTools MCP
description: Let agents inspect your site and WebView2 app with Chrome DevTools Model-Context-Protocol (MCP).
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: article
ms.service: microsoft-edge
ms.subservice: devtools
ms.date: 06/04/2026
---
# Let agents inspect your site and WebView2 app with Chrome DevTools MCP

Chrome DevTools for agents (`chrome-devtools-mcp`) lets your coding agent (such as Copilot, Antigravity, Claude, or Cursor) control and inspect a live Chromium-based browser, including Microsoft Edge and WebView2.

**Detailed contents:**
* [Introduction](#introduction)
* [Prerequisites](#prerequisites)
* [Launch Edge](#launch-edge)
   * [Additional flags for launching Edge](#additional-flags-for-launching-edge)
   * [Executable path for each Edge channel](#executable-path-for-each-edge-channel)

> * [Auto-connect to a running Edge instance](#auto-connect-to-a-running-edge-instance)

Reviewer 2: It might be worth considering if we should break out Auto-connect and Auto-connect to a WebView2 instance as separate docs or all part of the same one. I'll defer to what you recommend @mikehoffms

Author: https://review.learn.microsoft.com/microsoft-edge/web-platform/devtools-mcp-server?branch=pr-en-us-3811 --

<img width="670" height="321" alt="image" src="https://github.com/user-attachments/assets/67c4051a-ef9b-438e-864d-90086b9ceec1" />

<img width="515" height="783" alt="image" src="https://github.com/user-attachments/assets/da741e43-2bc1-45dc-85a9-3784abcf833d" />

Author: consider 3 articles instead of 1:

Article 1: **Let agents inspect your site and WebView2 app with Chrome DevTools MCP**
  * [Introduction](#introduction)
  * [Prerequisites](#prerequisites)
  * [Launch Edge](#launch-edge)
  * [How it works](#how-it-works)
  * [Configuring other MCP clients](#configuring-other-mcp-clients)
  * [Troubleshooting](#troubleshooting)
  * [See also](#see-also)

Article 2: **Auto-connect to a running Edge instance**
* [Step 1: Enable remote debugging in Edge](#step-1-enable-remote-debugging-in-edge)
* [Step 2: Configure the MCP server](#step-2-configure-the-mcp-server)
* [Step 3: Test your setup](#step-3-test-your-setup)
* [User data directory for each Edge channel](#user-data-directory-for-each-edge-channel)

Article 3: **Auto-connect to a WebView2 instance**
* [Step 1: Enable remote debugging for WebView2](#step-1-enable-remote-debugging-for-webview2)
   * [By using WebView2Utilities](#by-using-webview2utilities)
   * [By using the Windows Registry](#by-using-the-windows-registry)
* [Step 2: Find the WebView2 user data directory](#step-2-find-the-webview2-user-data-directory)
* [Step 3: Configure the MCP server](#step-3-configure-the-mcp-server)
* [Step 4: Test your setup](#step-4-test-your-setup)

@captainbrosset @zoherghadyali

Author: Docs team decision about doc-design: keep as 1 article, not 3.

@captainbrosset @zoherghadyali

| Channel | Default Path |
|---|---|
| Stable | `/usr/bin/microsoft-edge` |
| Beta | `/usr/bin/microsoft-edge-beta` |
| Dev | `/usr/bin/microsoft-edge-dev` |
| Canary | n/a; there isn't a Canary channel for Linux. |

---

<!-- ====================================================================== -->
## Auto-connect to a running Edge instance

Use the configuration below to connect to an already running Edge instance.  This is useful when:

* You want to maintain browser state between manual testing and agent-driven testing.
* You need to be signed into a website that blocks automated browser launches.
* You want to inspect an existing session.

> [!WARNING]
> When using auto-connect, your agent inherits your active session, including logged-in accounts, cookies, and other data surfaced through JavaScript APIs.  This may expose Personally Identifiable Information (PII) to the agent.  Only use this mode with agents you trust and be careful with your prompts.

<!-- ------------------------------ -->
#### Step 1: Enable remote debugging in Edge

There are have two options:

**Option A:** Start Edge with the remote debugging flag:

`msedge.exe --remote-debugging-port=9222`

> **Option B:** In a running Edge instance, navigate to `edge://inspect/#remote-debugging` and enable remote debugging from that page.

Reviewer 2: Might be worth including a screenshot of the page and explicit instructions to click/enable the "Allow remote debugging for this browser instance" checkbox.

<img width="2496" height="1592" alt="Image" src="https://github.com/user-attachments/assets/f014a33d-255a-43d3-804b-267bd65e2ee6" />

Author: <img width="903" height="457" alt="image" src="https://github.com/user-attachments/assets/eec3e3b5-4e54-4d0e-b6ad-c6f90486cb66" />

<img width="754" height="448" alt="image" src="https://github.com/user-attachments/assets/79d73801-eceb-4a1a-86bb-eff21f9f5e2e" />

---
title: Let agents inspect your site and WebView2 app with Chrome DevTools MCP
description: Let agents inspect your site and WebView2 app with Chrome DevTools Model-Context-Protocol (MCP).
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: article
ms.service: microsoft-edge
ms.subservice: devtools

> ms.date: 06/04/2026

Reviewer: ```suggestion
ms.date: 06/10/2026
```

---
# Let agents inspect your site and WebView2 app with Chrome DevTools MCP

Chrome DevTools for agents (`chrome-devtools-mcp`) lets your coding agent (such as Copilot, Antigravity, Claude, or Cursor) control and inspect a live Chromium-based browser, including Microsoft Edge and WebView2.

> **Detailed contents:**

Reviewer: ```suggestion
**Detailed contents:**

```
